### PR TITLE
Add login with custom token

### PR DESCRIPTION
**Pull request by cheungpat**
connects #153

### CODE REVIEW

File: Pod/Extensions/SSO/SKYAuthContainer+SSO.m [hunk 1]
```diff
@@ -161,4 +164,29 @@ - (NSURL *)sso_genCallbackURL:(NSString *)scheme
...
+    op.loginCompletionBlock = ^(SKYRecord *user, SKYAccessToken *accessToken, NSError *error) {
+        if (error) {
+            if (completionHandler) {
+                completionHandler(user, error);
+            }
+            return;
+        }
+
+        [self.container.auth updateWithUser:user accessToken:accessToken];
+        [self.container.push registerDeviceCompletionHandler:nil];
```
Comment (carmenlau) [inline Pod/Extensions/SSO/SKYAuthContainer+SSO.m]: I moved `registerDeviceCompletionHandler` inside `updateWithUser` in the last PR. So we can remove it here. The other parts look good!

refs: https://github.com/SkygearIO/skygear-SDK-iOS/blob/3e58465b9d6d99b78c90b2004821423908cd1303/Pod/Classes/SKYAuthContainer.m#L187

**Comment by cheungpat** [comment]
updated

thanks @carmenlau

File: Pod/Extensions/SSO/SKYAuthContainer+SSO.m [hunk 2]
```diff
@@ -161,4 +164,28 @@ - (NSURL *)sso_genCallbackURL:(NSString *)scheme
...
+                completionHandler(user, error);
+            }
+            return;
+        }
+
+        [self.container.auth updateWithUser:user accessToken:accessToken];
+
+        if (completionHandler) {
+            completionHandler(user, error);
+        }
```
Comment (carmenlau) [inline Pod/Extensions/SSO/SKYAuthContainer+SSO.m]: @cheungpat I just tried the api and do some ui operation in the `completionHandler` and got error `UI API called on a background thread`. I think we also need to call the `completionHandler` in main thread like the other login operation? Both error and success handler.

https://github.com/SkygearIO/skygear-SDK-iOS/blob/54b3f0d0475f3b14290842a2ffd5c1aca7111bb7/Pod/Classes/SKYAuthContainer.m#L94

thanks kuen kuen!!

Comment (cheungpat) [reply Pod/Extensions/SSO/SKYAuthContainer+SSO.m]: right

**Comment by cheungpat** [comment]
@carmenlau updated

File: Pod/Extensions/SSO/SKYAuthContainer+SSO.m [hunk 3]
```diff
@@ -161,4 +164,30 @@ - (NSURL *)sso_genCallbackURL:(NSString *)scheme
...
+
+- (void)loginWithCustomToken:(NSString *)customToken
+           completionHandler:(SKYContainerUserOperationActionCompletion)completionHandler
+{
+    SKYLoginCustomTokenOperation *op =
+        [SKYLoginCustomTokenOperation operationWithCustomToken:customToken];
+    op.loginCompletionBlock = ^(SKYRecord *user, SKYAccessToken *accessToken, NSError *error) {
+        if (error) {
+            if (completionHandler) {
+                completionHandler(user, error);
```
Comment (carmenlau) [inline Pod/Extensions/SSO/SKYAuthContainer+SSO.m]: I think we also need that here, running in main thread. after updating this, we can merge. :P

**Comment by cheungpat** [comment]
@carmenlau updated

**Pull request merged**